### ISSUE TITLE
Invalid value: "null" when field is empty in configuration and SSA is used to apply

### ISSUE THREAD

**Issue by karunasagark** [kubernetes/kubernetes#104694]
#### What happened:

We are developing a new k8s controller and decided to use SSA completely. We ran into an issue when I marked a field empty in the configuration for an existing object. I have a custom resource with field called 'prop1' and it is not a required field (see snippet below). 

Scenario 1:
Using SSA, I created a new object of this CR kind with 'prop1' specified and this was successful. I changed the configuration by marking it empty i.e. 'prop1: {}' and used the same code aka SSA to apply this configuration, but this time I received an error - Invalid value: "null": spec.prop1 in body must be of type object: "null".

Scenario 2:
Using SSA, I created a new object of this CR kind with 'prop1: {}' and using the SSA to apply. This worked without errors and the 'get' of object shows that the field was empty.

So, from above scenario, you can see the inconsistency between SSA apply on new object vs existing object when the field is marked empty.

This is closely related to https://github.com/kubernetes/kubernetes/issues/103011, expect in this case instead of removing the field from the configuration we are marking it as empty.

```
spec:
            type: object
            properties:
              prop1:
                type: object
                properties:
                  enabled:
                    type: boolean
              prop2:
                type: object
                properties:
                  enabled:
                    type: boolean
```

#### What you expected to happen:

Was not expecting this error instead was expecting the field to be marked empty.

Can you please help me understand this behavior?

#### How to reproduce it (as minimally and precisely as possible):

Attached zip file containing testcrd.yaml, testcr1.json, testcr2.json and testcr3.json

[reproJson.zip](https://github.com/kubernetes/kubernetes/files/7086582/reproJson.zip)

##### Scenario 1
```
$ kubectl apply --server-side --force-conflicts -f testcrd.yaml
customresourcedefinition.apiextensions.k8s.io/testones.example.com serverside-applied


$ kubectl apply --server-side --force-conflicts -f testcr1.json
testone.example.com/objectone serverside-applied


$ kubectl apply --server-side --force-conflicts -f testcr2.json
The TestOne "objectone" is invalid: spec.prop1: Invalid value: "null": spec.prop1 in body must be of type object: "null"

$ diff testcr1.json testcr2.json
9,11c9
<       "prop1": {
<           "enabled": false
<       },
---
>       "prop1": {},


$ kubectl get testone objectone -o json --show-managed-fields
{
    "apiVersion": "example.com/v1beta1",
    "kind": "TestOne",
    "metadata": {
        "creationTimestamp": "2021-08-31T20:05:31Z",
        "generation": 1,
        "managedFields": [
            {
                "apiVersion": "example.com/v1beta1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:spec": {
                        "f:prop1": {
                            "f:enabled": {}
                        },
                        "f:prop2": {
                            "f:enabled": {}
                        }
                    }
                },
                "manager": "kubectl",
                "operation": "Apply",
                "time": "2021-08-31T20:05:31Z"
            }
        ],
        "name": "objectone",
        "namespace": "default",
        "resourceVersion": "6777340",
        "uid": "78b9e081-436b-4260-aa93-adaf9ae69c06"
    },
    "spec": {
        "prop1": {
            "enabled": false
        },
        "prop2": {
            "enabled": false
        }
    }
}
```

##### Scenario 2
```
$ diff testcr2.json testcr3.json
5c5
<         "name": "objectone",
---
>         "name": "objecttwo",


$ kubectl apply --server-side --force-conflicts -f testcr3.json
testone.example.com/objecttwo serverside-applied


$ kubectl get testone objecttwo -o json --show-managed-fields
{
    "apiVersion": "example.com/v1beta1",
    "kind": "TestOne",
    "metadata": {
        "creationTimestamp": "2021-08-31T20:07:12Z",
        "generation": 1,
        "managedFields": [
            {
                "apiVersion": "example.com/v1beta1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:spec": {
                        "f:prop1": {},
                        "f:prop2": {
                            "f:enabled": {}
                        }
                    }
                },
                "manager": "kubectl",
                "operation": "Apply",
                "time": "2021-08-31T20:07:12Z"
            }
        ],
        "name": "objecttwo",
        "namespace": "default",
        "resourceVersion": "6777495",
        "uid": "ff3a878a-d91a-4aac-b1ff-b52f9412bb36"
    },
    "spec": {
        "prop1": {},
        "prop2": {
            "enabled": false
        }
    }
}
```

#### Anything else we need to know?:

- CCA was NOT used
- Seems closely related to https://github.com/kubernetes/kubernetes/issues/103011
- Ofcourse, we can add 'nullable: true' to the CRD and it does help. However, I'm trying to understand where 'null' is coming from and trying to see if we get consistency when SSA is applied to new or existing objects.

#### Environment:

- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.7", GitCommit:"4966fd0dded6419fd23233a38c8905c67a3ea78e", GitTreeState:"clean", BuildDate:"2021-08-17T02:06:07Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
Azure Kubernetes Service

- OS (e.g: `cat /etc/os-release`):
Ubuntu 18.04

- Kernel (e.g. `uname -a`):

- Install tools:

- Network plugin and version (if this is a network-related bug):

- Others:

**Comment by jpbetz** [comment]
I suspect this will also be fixed by https://github.com/kubernetes-sigs/structured-merge-diff/pull/184 (xref https://github.com/kubernetes/kubernetes/issues/103011#issuecomment-866332473), but it would be good to validate that before closing this.

**Comment by karunasagark** [comment]
@jpbetz from this [comment](https://github.com/kubernetes/kubernetes/issues/103011#issuecomment-866332473), it seems that fix is included in 1.20.7. However, the above repro steps were performed on 1.20.7, so there seems to be something missing?

**Comment by jpbetz** [comment]
> Joe Betz from this [comment](https://github.com/kubernetes/kubernetes/issues/103011#issuecomment-866332473), it seems that fix is included in 1.20.7. However, the above repro steps were performed on 1.20.7, so there seems to be something missing?

Thanks for checking. I'll look more closely at this.

**Comment by chhsia0** [comment]
This issue is caused by https://github.com/kubernetes-sigs/structured-merge-diff/pull/259

/reopen
